### ISSUE TITLE
OOM when switching to distributed training

### ISSUE THREAD

**Issue by VictorSanh** [huggingface/accelerate#213]
I am working on an [evaluation script](https://github.com/bigscience-workshop/t-zero/blob/59acfb46124f44b25559aee7862f93e135b80e90/evaluation/run_eval.py) which was adapted from [transformers/examples/pytorch/multiple-choice/run_swag_no_trainer.py](https://github.com/huggingface/transformers/blob/7533d30acd975027e83a548e4c38e06fa335291b/examples/pytorch/multiple-choice/run_swag_no_trainer.py).

I developed it using a single node/single GPU:
```bash
python run_eval.py \
    --dataset_name super_glue \
    --dataset_config_name rte \
    --template_name "must be true" \
    --model_name_or_path bigscience/T0_3B \
    --output_dir ./debug
```
and it works perfectly fine on a 24GB GPU (topping at 14GB) with a default batch size of 8.

As soon as I use distributed training (`python -m torch.distributed.launch --nproc_per_node 2 ...` or `accelerate launch ...` after configuring with `accelerate config`), I run into OOM errors, the best I can fit on each 24GB device is batch size 2...

Any idea what could be happening here?

**Comment by sgugger** [comment]
I'm not able to fully reproduce as using distributed training on my machine hangs at the model loading (probably due to a lack of RAM).

Trying with a smaller model, I can see there is a bit more usage of GPU memory when going distributed indeed, and I think the cause is that PyTorch resevers memory for the buffers that synchronize the gradients, even if they end up being not needed here. So if the script only does eval, just do not send `model` to the `accelerator.prepare` method and just do:
```
    # Prepare everything with our `accelerator`.
    eval_dataloader = accelerator.prepare(eval_dataloader)
```
in your script. Using this way, I ended up with the same memory usage launching with python or accelerate launch.

**Comment by VictorSanh** [comment]
thank you @sgugger ! it works as intended now 🎉
closing now
